### PR TITLE
New File Submission Service For Share Extension

### PR DESCRIPTION
**Pull request by vargaat**
## Checklist

- [x] Follow-up e2e test ticket created or not needed
- [x] A11y checked
- [x] Tested on phone
- [x] Tested on tablet
- [x] Approve from product or not needed

refs: MBL-16145
affects: Student
release note: none
test plan: Try to break file upload when submitting via file extension submission. Background upload is only supported when the dismiss button appears during upload.

### CODE REVIEW

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Build failed, skipping coverage check
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  This pull request will not generate a release note.
  </td>
    </tr>
  </tbody>
</table>


<details><summary>:x: XCTest failed: CoreTests/BackgroundURLSessionProviderTests/testCreatesNewSessionIfSessionBecameInvalid</summary>


    Assertion Failure at BackgroundURLSessionProviderTests.swift:62: XCTAssertNotEqual failed: ("<__NSURLBackgroundSession: 0x7f9c51ab8300>") is equal to ("<__NSURLBackgroundSession: 0x7f9c51ab8300>")
</details>


[MBL-16145](https://instructure.atlassian.net/browse/MBL-16145)

#### Affected Apps: Student
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: Core/CoreTests/Files/FileSubmission/Model/BinaryUpload/BackgroundURLSessionProviderTests.swift [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+class BackgroundURLSessionProviderTests: CoreTestCase {
+    private var testee: BackgroundURLSessionProvider!
+    private var mockUploadProgressObserversCache: MockFileUploadProgressObserversCache!
+
+    override func setUp() {
+        super.setUp()
+        mockUploadProgressObserversCache = MockFileUploadProgressObserversCache()
+        testee = BackgroundURLSessionProvider(sessionID: "testSession", sharedContainerID: "testContainer", uploadProgressObserversCache: mockUploadProgressObserversCache)
+    }
```
Comment (szabolcs-instructure) [inline Core/CoreTests/Files/FileSubmission/Model/BinaryUpload/BackgroundURLSessionProviderTests.swift:31]: Just to be super safe maybe we could set `testee` and `mockUploadProgressObserversCache` to nil in the `tearDown` function

**Review by szabolcs-instructure — APPROVED** [review]
QA + 1
I mostly had cosmetic remarks only

File: Core/Core/Files/FileSubmission/Model/BinaryUpload/FileUploadProgressObserversCache.swift [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            let url = URL(string: stringURL)
+        else { return nil }
+
+        return context.persistentStoreCoordinator?.managedObjectID(forURIRepresentation: url)
+    }
+
+    /**
+     - returns: The `FileUploadProgressObserver` for the given upload item either from the cache or a newly created instance.
+     */
+    public func retrieveProgressObserver(fileUploadItemID: NSManagedObjectID, fileSubmissionID: NSManagedObjectID) -> FileUploadProgressObserver {
```
[the remark this answers is not in the document]
Comment (vargaat) [reply Core/Core/Files/FileSubmission/Model/BinaryUpload/FileUploadProgressObserversCache.swift:60]: I also made `makeObjectID` function private.

File: Core/Core/Files/FileSubmission/Model/BinaryUpload/FileUploadProgressObserversCache.swift [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            return observer
+        }
+
+        let observer = factory(fileSubmissionID, fileUploadItemID)
+        removeObserverOnCompletion(observer)
+        observerCache[fileUploadItemID] = observer
+        return observer
+    }
+
+    private func retrieveProgressObserver(taskID: String?) -> FileUploadProgressObserver? {
```
[the remark this answers is not in the document]
Comment (vargaat) [reply Core/Core/Files/FileSubmission/Model/BinaryUpload/FileUploadProgressObserversCache.swift]: Refactored.

**Review by idevainstr — CHANGES_REQUESTED** [review]
just some minors...

File: Core/Core/Files/FileSubmission/Model/BinaryUpload/FileUploadProgressObserver.swift [hunk 4]
```diff
@@ -51,13 +52,27 @@ extension FileUploadProgressObserver: URLSessionTaskDelegate {
...
+        if error?.isBackgroundSessionDisconnected == true {
+            completionSubject.send(completion: .failure(.uploadContinuedInApp))
+            return
+        }
 
-            if item.apiID == nil, error == nil {
-                item.uploadError = NSLocalizedString("Session completed without error or file ID.", comment: "")
+        context.performAndWait { [weak self] in
+            guard let item = try? context.existingObject(with: fileUploadItemID) as? FileUploadItem,
+                  let self = self
```
[the remark this answers is not in the document]
Comment (vargaat) [reply Core/Core/Files/FileSubmission/Model/BinaryUpload/FileUploadProgressObserver.swift]: Actually this `performAndWait` accepts a non-escaping block so we don't need this `[weak self]` at all.

File: Core/Core/Files/FileSubmission/Model/Entities/FileSubmissionErrors.swift [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+public struct FileSubmissionErrors {
+    public enum CoreData: Error, Equatable {
+        case submissionNotFound
+        case uploadItemNotFound
+    }
+    public enum UploadFinishedCheck: Error, Equatable {
+        case notReady
```
[the remark this answers is not in the document]
Comment (vargaat) [reply Core/Core/Files/FileSubmission/Model/Entities/FileSubmissionErrors.swift]: Agree, renamed.

**Review by szabolcs-instructure — APPROVED** [review]

**Review by idevainstr — APPROVED** [review]
QA +1 🏅

**Review by davidvarga93 — APPROVED** [review]
QA + 1

**Review by balintbartok — APPROVED** [review]
QA + 1

**Pull request merged**